src/searchbar.vue

```diff
@@ -0,0 +1,63 @@
+<script>
+import Typeahead from './Typeahead.vue';
+
+export default {
+  extends: Typeahead,
+  computed: {
+    primitiveData() {
+      if (this.data) {
+        const matches = [];
+        this.data.forEach((value) => {
+          const query = new RegExp(this.value, 'i');
+          let hasMatchingHeading = false;
+          const { title, headings, src } = value;
+          const keywords = value.keywords || '';
+          if (Object.keys(headings).length !== 0) {
+            Object.entries(headings).forEach(([id, text]) => {
+              if (text.search(query) !== -1) {
+                hasMatchingHeading = true;
+                matches.push({
+                  title,
+                  keywords,
+                  src,
+                  heading: { id, text },
+                });
+              }
+            });
+            if (!hasMatchingHeading) {
+              const words = keywords ? keywords + title : title;
+              if (words.search(query) !== -1) {
+                matches.push(value);
+              }
+            }
+          }
+          return {};
+        });
+
+        return matches;
+      }
+      return {};
```

Comment:
Refactoring `primitiveData()`.

Iteration 1:
- Early `return`.
- Rename `value` to `entry` as there's `this.value`.
- Rename `query` to `regex` as there's `this.query`.
- Reorder destructuring of `entry` alphabetically.
- Reorder `const` before `let`.
- Remove unnecessary `if (Object.keys(headings).length !== 0)`.
- Use `regex.test(text)` instead of `text.search(query) !== -1`.
- Reorder properties of object in `matches.push(object)` alphabetically.
- Replaced `words.search(query) !== -1` with `regex.test(title) || regex.test(keywords)`.

```js
if (!this.data) {
  return {};
}
const matches = [];
this.data.forEach((entry) => {
  const { headings, src, title } = entry;
  const keywords = entry.keywords || '';
  const regex = new RegExp(this.value, 'i');
  let hasMatchingHeading = false;
  Object.entries(headings).forEach(([id, text]) => {
    if (regex.test(text)) {
      hasMatchingHeading = true;
      matches.push({
        heading: { id, text },
        keywords,
        src,
        title,
      });
    }
  });
  if (!hasMatchingHeading) {
    if (regex.test(title) || regex.test(keywords)) {
      matches.push(entry);
    }
  }
});
return matches;
```

Now trying [`Array.prototype.reduce()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).

Reply:
Iteration 2:
- Return early with `undefined`.
- Assign `regex` outside `forEach`.

```js
if (!this.data) {
  return undefined;
}
const matches = [];
const regex = new RegExp(this.value, 'i');
this.data.forEach((entry) => {
  const { headings, src, title } = entry;
  const keywords = entry.keywords || '';
  let hasMatchingHeading = false;
  Object.entries(headings).forEach(([id, text]) => {
    if (regex.test(text)) {
      hasMatchingHeading = true;
      matches.push({
        heading: { id, text },
        keywords,
        src,
        title,
      });
    }
  });
  if (!hasMatchingHeading) {
    if (regex.test(title) || regex.test(keywords)) {
      matches.push(entry);
    }
  }
});
return matches;
```

Reply:
An attempt at `reduce` that's less appealing:

```js
if (!this.data) {
  return undefined;
}
const regex = new RegExp(this.value, 'i');
return this.data.reduce((matches, entry) => {
  const { headings, src, title } = entry;
  const keywords = entry.keywords || '';
  const headingMatches = Object.entries(headings).reduce((acc, [id, text]) => {
    if (regex.test(text)) {
      acc.push({
        heading: { id, text },
        keywords,
        src,
        title,
      });
    }
    return acc;
  }, []);
  if (headingMatches.length) {
    return matches.concat(headingMatches);
  }
  if (regex.test(title) || regex.test(keywords)) {
    matches.push(entry);
  }
  return matches;
}, []);
```

Reply:
@nusjzx Iteration 2 is more appealing.